### ISSUE TITLE
Drag & Drop Board For Memos (Memos 2.0/Big Bon Board 0.1.0)

### ISSUE THREAD

**Issue by Quorafind** [Quorafind/Obsidian-Thino#66]
- [x] Infinity Board
- [x] Allow users to drag & drop
- [x] Allow users to edit memo in board 
- [x] Allow users to line to each other
- [ ] Allow users to exported whole board as image
- [x] Support memos magnetic
- [x] Allow users to create memo in board
- [x] Allow users to drag and drop png\web link\video(?) in board
- [ ] Users to Manage all new memos in board to memos basic
- [ ] Users to Manage memos in board
- [ ] Allow users to drag and drop memo from basic to board
- [ ] Allow users to set default folder for memos
- [ ] Allow users transfer Memos Board to normal markdown document like: Kanban
- [ ] Allow users to drag memos from memos list or another memos board
- [x] Allow users to use hotkey to scroll hori\verti\zoom\pan
- [ ] Allow users to open memos in board to be document 
- [ ] Render Markdown by using Obsidian default renderer to make codeblock and excalidraw work.
- [ ] Render Excalidraw
- [ ] Allow use timeline to fetch cards from different period

**Comment by Quorafind** [comment]
@wenlzhang Need some ideas for design this feature. Now if you can drag and drop memos into a board that you can manage their arrangement(imagine excalidraw), what would you like it has except these tasks I have added above?

**Comment by wenlzhang** [comment]
This can be an exciting feature for Memos! 😃 I have some rough ideas regarding this feature, and I am also thinking about how to make this feature distinguish Memos from other similar plugins, e.g. Excalidraw and those for creating mind maps.

- This Big Bon Board itself can be a Markdown file, and it can consist of two sections, like the Excalidraw plugin.
    - The text section contains mainly the text and block links, which would be described later; and the code, e.g. `json`, section can be used to denote the relationships among Memos/nodes.
    - This way, one can easily link to anything in the text section from other notes, and such a relationship would also show up in the Backlinks section and Graph View of Obsidian.
- I imagine (current idea) that there can be three types of nodes that can be inserted into the board.
    - The first type (Type A) is Memos dragged from the Memos UI.
    - The second type (Type B) is thoughts created directly on the board.
    - The third type (Type C) is links to any other note in Obsidian.
- When dragging Type A Memos from the Memos UI into the Board, there can be different ways of doing this.
    - The first way is that there can be some kind of linking relationship with its original location in the daily notes, such that when editing the node on the board, its original text in the corresponding daily note would also change; the second way is that the note is copied from the daily note into the board, such that the change of the node on the board would not affect its original content.
    - Either way, when dragging Type A nodes from Memos UI into the Board, there should be a block link at the end, denoting its original location in the corresponding daily notes.
    - Personally, I would like the second way more, as I would like to keep its original content as it is for reference. Therefore, there can be an option in the configuration pane for this.
- When creating Type B nodes, there should be a time stamp inserted at the end (or other locations) of the node, denoting when the nodes are created.
    - At anytime, one can choose to insert Type B nodes into daily notes, just like when creating ordinary Memos.
    - The date and time when the insertion happens would determine which daily note this node should be inserted to.
    - One can insert such nodes back into Memos by dragging the node from the Board into the Memos UI, by using right-click menu or by some kind of hotkeys.
    - When inserting such nodes into daily notes, there should be a block link to its original location to the Board note.

Example of Big Bon Board:
```
---

memos-mode: big-bon-board

---

# Big Bon Board

## Text

This is a Type A node. [[2022-01-21#^BlockIDA]]

This is a Type B node. 📝 2022-01-21T12:00 ^BlockIDB

The following are Type C nodes:

[[Note A]]
![[Note B#Header A]]
![[Note C#^BlockIDA]]

## Code



```

Example Daily Note with inserted node from the Board:
```

# 2022-06-01

## Memos

- 13:00 This is a Type B node. 📝 2022-01-21T12:00 [[Big Bon Board#^BlockIDB]]

```

**Comment by Quorafind** [comment]
Very useful suggestions, I am building a poc work for this. If I am on vocation, it should come in the Feb.

**Comment by wenlzhang** [comment]
Happy Chinese new year to you in advance!

This feature of Memos reminds me of another tool, Scapple, which can be used to draft and connect ideas.
https://www.literatureandlatte.com/scapple/overview

I feel Memos can be more powerful and flexible than that, considering the power of Obsidian and its plugins. I have a few other thoughts on this feature:

- Besides Type B nodes, whenever a node is created, there can be a timestamp attached. Therefore, the example above can become something like the following.
    - Although such timestamps would be created in the corresponding Markdown file of the Board, it could be that one may not want to visualize them on the board. Therefore, there can be an option (on the board or in the configuration pane) for choosing whether to show such timestamps on the board.
- Besides text and note links, it would be great if one can also embed images, Excalidrawings, MarkMind mind maps, etc.
- If an element is deleted from the Board, there would be records for this in files like `delete.md`, just like when deleting Memos.

```
---

memos-mode: big-bon-board

---

# Big Bon Board

## Text

This is a Type A node. 📝 2022-01-21T10:00 [[2022-01-21#^BlockIDA]]

This is a Type B node. 📝 2022-01-21T12:00 ^BlockIDB

The following are Type C nodes:

[[Note A]] 📝 2022-01-21T13:00
![[Note B#Header A]] 📝 2022-01-21T14:00
![[Note C#^BlockIDA]] 📝 2022-01-21T15:00

## Code



```

**Comment by Quorafind** [comment]
How about when you can manage all the board in Memos, and then all the boards links(like card view) live in a root board(named memos.board.md?), and then you can create a sub board in the root board, and then, you can select and dragNdrop memo into the sub board(of course, you can create memo in any subboard(All these willsend to daily note like memos do)).

**Comment by wenlzhang** [comment]
> How about when you can manage all the board in Memos, and then all the boards links(like card view) live in a root board(named memos.board.md?), and then you can create a sub board in the root board

I think this is a great idea! This way, one can easily keep track of all boards created with Memos.

Does this mean that all sub-boards would be automatically inserted/linked into the root board? Is it possible to also support the following for this root board?

- One can group the sub-boards on the root board in some way so that one can easily manage all sub-boards.
    - For instance, some sub-boards would be placed under the same header of `memos.board.md`, e.g. `## Topic 1`.
    - Alternatively, one can assign tags to sub-boards and such tags can show up in the Memos UI or somewhere in `memos.board.md` close to the sub-board name, or maybe inside the sub-board Markdown file.
- One can add brief description/comment to sub-boards on the root board.
    - Such a description/comment can be an item of another attribute, or an ordinary text/memo.

This feature of Memos reminds me of another tool, Heptabase: https://heptabase.com/

**Comment by Quorafind** [comment]
Yes, you can refer what heptabase has done. I am thinking of a good way to make things as robust as I can. Using ![[xxxxxxx.md^123412]] should be ok.

But on the other hand, if we could use memo as a block, and even create a sub-board for it, it should be cool.

**Comment by wenlzhang** [comment]
> But on the other hand, if we could use memo as a block, and even create a sub-board for it, it should be cool.

Definitely! Can't wait to try it!

**Comment by Quorafind** [comment]
![image](https://user-images.githubusercontent.com/13215013/153355149-64dd479c-0b10-4de6-8135-ed07ba89a284.png)

As reference.

**Comment by Quorafind** [comment]
![Memos 2 0 (1)_1_26](https://user-images.githubusercontent.com/13215013/155935755-81c44560-182a-4154-80c7-d1ff4e3e0388.png)

Update @wenlzhang , hope you love the interface.

**Comment by wenlzhang** [comment]
It looks really impressive! Would it be possible to add annotations/text to lines?

**Comment by Quorafind** [comment]
![image](https://user-images.githubusercontent.com/13215013/155950110-6a72cb8a-16dc-49ca-8196-7e4e5b617b0f.png)

Yes, annotations and Readwise support features will be released in next version @wenlzhang

**Comment by Quorafind** [comment]
![Memos 设计稿 (1)_3_26](https://user-images.githubusercontent.com/13215013/155950362-c1da9ee7-4ac7-428b-9169-fb85607f6174.png)

![Memos 设计稿 (1)_1_26](https://user-images.githubusercontent.com/13215013/155950421-934cb538-796b-42c2-96a8-11a71c3eb140.png)

![Memos 设计稿 (1)_2_26](https://user-images.githubusercontent.com/13215013/155950427-0fa189be-fd57-4a9f-b1b6-2c42e18350c4.png)

Yes, annotations feature will be released in 1.9.0

**Comment by red-co** [comment]
> ![备忘录 2 0 (1)_1_26](https://user-images.githubusercontent.com/13215013/155935755-81c44560-182a-4154-80c7-d1ff4e3e0388.png)
> 

- 能允许调整板子的大小吗？
 - 大小会在视觉上给出一个质量感的区分，
 - 在没认真看内容之前，你就能直观的意识到一些比一些更重要，而某一些比其它的更不重要，
非常感谢你制作的插件，(｡･ω･｡)ﾉ♡
translate to en :
- Is it possible to resize the board?
- Size will visually give a sense of quality distinction,
- Before you look at the content carefully, you can intuitively realize that some are more important than some, and some are less important than others

**Comment by red-co** [comment]
这是我对这个插件的一点理解，(希望我没有对这个插件误会什么，😄，)
---
excalidraw和memo它们两个的目的和倾向不同 
1. excalidraw，它就是白板，在这上面写东西，思考，在这个上面创建元素，修改，调整，绘图，
它会让我们在这个绘图页面上面写更多，调整的更多，但是对obsidian内其它的文档的支持不太好，(历史遗留原因，因为这个程序设计之初的目的就只是白板，用于会议的，，而不是我们这样的拿来进行持续的话题进行分析，相互引用，)
2. memos(hetp,)，更倾向于引用和管理原有的内容，
- 它和obsidian的设计目的很像，不对文本和元素本身负责，而是给了我们一种操作文档的更直觉的交互方式，
- 它会让我们倾向于在这个上面更好的联系它周围的东西，比如另一个板子，或者obsidian内某处的内容，
因为它可以双向的对内容进行修改，(至少这是它的目的，)，而excalidraw对文件间的交互支持不太好，因为这本来就不是它的目的，

**Comment by wenlzhang** [comment]
Can't wait to test it!

**Comment by wenlzhang** [comment]
Thanks for the efforts on developing Memos and this feature! Just bought you a few cups of coffee.

**Comment by wolfHts** [comment]
期待2.0版本的到来！

**Comment by wenlzhang** [comment]
Hope everything goes well with you! Is there any update with the development of this board feature?

**Comment by Quorafind** [comment]
Not yet, I am working on other projects now for polishing my coding skill. And I will update memos in October

**Comment by jzhang-brex** [comment]
Obsidian just release the Canvas feature. It looks very similar to this one. :)

**Comment by Quorafind** [comment]
Now you can use Thino with canvas features.

**Comment by Quorafind** [comment]
Available in 2.3.43 for Pro users.
